src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/ProjectNames.cs

```diff
@@ -103,11 +103,73 @@ public static async Task<ProjectNames> FromDTEProjectAsync(EnvDTE.Project dtePro
                 projectId: projectId);
         }
 
-        private static string GetProjectGuid(string fullname, IVsSolution5 vsSolution5)
+        private static string GetProjectGuid(EnvDTE.Project dteProject, SVsSolution vsSolution)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-            var guid = vsSolution5.GetGuidOfProjectFile(fullname);
-            return guid.ToString();
+
+            // Each project system needs to implement its own DTE project implementation. Hence, custom project types
+            // have the potential to not implement everything. Therefore, we need to try multiple ways to maximise the
+            // chance we find one way that actually works.
+
+            if (TryGetProjectGuidFromUniqueName(dteProject, (IVsSolution2)vsSolution, out Guid guid))
+            {
+                return guid.ToString();
+            }
+
+            if (TryGetProjectGuidFromFullName(dteProject, (IVsSolution5)vsSolution, out guid))
+            {
+                return guid.ToString();
+            }
+
+            throw new ArgumentException("Unable to find project guid");
+
+            bool TryGetProjectGuidFromUniqueName(EnvDTE.Project project, IVsSolution2 vsSolution2, out Guid projectGuid)
+            {
+                try
+                {
+                    var uniqueName = project.UniqueName;
+                    if (string.IsNullOrEmpty(uniqueName))
+                    {
+                        projectGuid = default(Guid);
+                        return false;
+                    }
+
+                    ErrorHandler.ThrowOnFailure(vsSolution2.GetProjectOfUniqueName(uniqueName, out IVsHierarchy hierarchy));
+                    ErrorHandler.ThrowOnFailure(vsSolution2.GetGuidOfProject(hierarchy, out projectGuid));
```

Comment:
I see from the docs that [`IVsSolution`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution?view=visualstudiosdk-2019) also provides an implementation to `GetProjectOfUniqueName` and `GetGuidOfProject` methods. I am curious to know what happens if a language implements this interface instead of `IVsSolution2` that is derived from `IVsSolution`.

Reply:
IVsSolution2 extends IVsSolution, so it automatically gets all the methods, hence IVsSolution2's GetGuidOfProject is actually from IVsSolution. I believe I didn't need to use IVsSolution2 at all, it could be swapped to the other interface. But, these interfaces are implemented by Visual Studio's SVsSolution, which I believe means that it is not up to project system implementations. Or if it is, it's SVsSolution's implementation responsible for calling into the project system. But again IVsSolution2 extends IVsSolution, so I'd expect them to have the same implementation. As far as I've heard, only the DTE interfaces are implemented by project systems.